inspirehep/dojson/hepnames/fields/bd1xx.py

```diff
@@ -159,21 +165,27 @@ def _guess_type_from_value(a_value):
         if INSPIRE_BAI.match(a_value):
             return 'INSPIRE BAI'
 
+    def _try_to_correct_value(_type, a_value):
+        if type_ == 'CERN' and LOOKS_LIKE_CERN.match(a_value):
+            return 'CERN-' + NON_DIGIT.sub('', a_value)
+        elif type_ == 'KAKEN':
+            return 'KAKEN-' + a_value
+        else:
+            return a_value
+
     a_value = force_single_element(value.get('a'))
 
     type_ = _get_type(value)
     if type_ is None:
         type_ = _guess_type_from_value(a_value)
```

Comment:
You think? The algorithm is:

- Get the type from `035__9`
- If it didn't work try to guess the type from the value in `035__a`
- Given the type and the value, try to massage the value so that it respects the type
- If you found both a type and a value, return them

The general idea is that we trust types more than values.